--- git/configure.ac	2016-11-13 02:01:31.927576997 +0100
+++ git/configure.ac.new	2016-11-13 02:10:18.263557439 +0100
@@ -144,25 +144,6 @@
 if test $ac_auto_cflags = yes; then
 	if test $ac_enable_debug = yes ; then
 		CFLAGS="-O0 -g -Wall -Wno-sign-compare -Wno-unused"
-	else
-		CFLAGS=""
-		dnl GCC has problems with -march=native with Raspberry,
-		dnl resulting in a "double free or corruption" error
-		dnl See: https://www.raspberrypi.org/forums/viewtopic.php?f=28&t=139091
-		if ! $GREP ARM /proc/cpuinfo >/dev/null 2>&1; then
-			AC_CHECK_CC_OPT([-march=native], [CFLAGS="$CFLAGS -march=native"], [])
-		fi
-		AC_CHECK_CC_OPT([-O2], [CFLAGS="$CFLAGS -O2"], [CFLAGS="$CFLAGS -O"])
-		AC_CHECK_CC_OPT([-fomit-frame-pointer], [CFLAGS="$CFLAGS -fomit-frame-pointer"], [])
-		dnl Code is not ready for strict-aliasing. It seems working, but who know...
-		AC_CHECK_CC_OPT([-fno-strict-aliasing], [CFLAGS="$CFLAGS -fno-strict-aliasing"], [])
-		dnl Code is not ready for strict-overflow. It seems working, but who know...
-		AC_CHECK_CC_OPT([-fno-strict-overflow], [CFLAGS="$CFLAGS -fno-strict-overflow"], [])
-		dnl We are not interested in stack protection
-		AC_CHECK_CC_OPT([-fno-stack-protector], [CFLAGS="$CFLAGS -fno-stack-protector"], [])
-		AC_CHECK_CC_OPT([-Wall], [CFLAGS="$CFLAGS -Wall"], [])
-		AC_CHECK_CC_OPT([-Wno-sign-compare], [CFLAGS="$CFLAGS -Wno-sign-compare"], [])
-		AC_CHECK_CC_OPT([-Wno-unused], [CFLAGS="$CFLAGS -Wno-unused"], [])
 	fi
 	if test ! -z $ac_host_cflags; then
 		CFLAGS="$ac_host_cflags $CFLAGS"
