diff -Nur git.old/Makefile git/Makefile
--- git.old/Makefile	2017-03-05 03:34:26.000000000 +0100
+++ git/Makefile	2017-03-05 03:34:58.585478950 +0100
@@ -3,10 +3,10 @@
 
 
 BINARY    ?= wolf3d
-PREFIX    ?= /usr/local
-MANPREFIX ?= $(PREFIX)/share/man/
-MANPAGE   ?= man6/wolf4sdl.6
-DATADIR   ?= $(PREFIX)/share/games/wolf3d/
+PREFIX    ?= /usr
+MANPREFIX ?= $(PREFIX)/share/man/
+MANPAGE   ?= man6/wolf4sdl.6
+DATADIR   ?= $(PREFIX)/share/games/wolf3d/
 
 INSTALL         ?= install
 INSTALL_PROGRAM ?= $(INSTALL) -m 555 -s
@@ -14,9 +14,9 @@
 INSTALL_DATA    ?= $(INSTALL) -m 444
 
 
-SDL_CONFIG  ?= sdl-config
-CFLAGS_SDL  ?= $(shell $(SDL_CONFIG) --cflags)
-LDFLAGS_SDL ?= $(shell $(SDL_CONFIG) --libs)
+SDL_CONFIG  ?= pkg-config
+CFLAGS_SDL  ?= $(shell $(SDL_CONFIG) sdl --cflags)
+LDFLAGS_SDL ?= $(shell $(SDL_CONFIG) sdl --libs)
 
 
 CFLAGS += $(CFLAGS_SDL)
@@ -29,10 +29,10 @@
 CFLAGS += -Wcast-align
 
 
-ifdef DATADIR
-    CFLAGS += -DDATADIR=\"$(DATADIR)\"
-endif
-
+ifdef DATADIR
+    CFLAGS += -DDATADIR=\"$(DATADIR)\"
+endif
+
 CCFLAGS += $(CFLAGS)
 CCFLAGS += -std=gnu99
 CCFLAGS += -Werror-implicit-function-declaration
@@ -117,4 +117,4 @@
 	@echo '===> INSTALL'
 	$(Q)$(INSTALL) -d $(PREFIX)/bin
 	$(Q)$(INSTALL_PROGRAM) $(BINARY) $(PREFIX)/bin
-	$(Q)$(INSTALL_MAN) $(MANPAGE) $(MANPREFIX)/man6
+	$(Q)$(INSTALL_MAN) $(MANPAGE) $(MANPREFIX)/man6
