--- git/Makefile.in	2017-03-29 15:30:33.812709327 +0200
+++ git/Makefile.in.new	2017-03-29 15:30:02.788709068 +0200
@@ -153,6 +153,7 @@
 $(objects)/$(SDLTEST_TARGET): $(SDLTEST_OBJECTS)
 	$(RUN_CMD_AR)$(AR) cru $@ $(SDLTEST_OBJECTS)
 	$(RUN_CMD_RANLIB)$(RANLIB) $@
+	$(RUN_CMD_LTLINK)$(LIBTOOL) --tag=CC --mode=link $(CC) -o $@ $(SDLTEST_OBJECTS) $(LDFLAGS) $(EXTRA_LDFLAGS) $(LT_LDFLAGS)
 
 install: all install-bin install-hdrs install-lib install-data
 install-bin:
