--- dgen-sdl-1.32/configure.ac	2013-02-09 21:26:18.000000000 +0100
+++ dgen-sdl-1.32.build/configure.ac	2017-02-01 22:58:22.129811634 +0100
@@ -13,7 +13,7 @@
 AC_CANONICAL_SYSTEM
 
 dnl Initialize Automake.
-AM_INIT_AUTOMAKE([silent-rules])
+AM_INIT_AUTOMAKE([subdir-objects])
 AM_SILENT_RULES([yes])
 AM_PROG_AS
 
@@ -110,7 +110,12 @@
 TOBIN_LIBS="${LIBS} ${TOBIN_LIBS}"
 
 dnl Check for SDL libs.
-AM_PATH_SDL([1.2.0], [], [AC_MSG_ERROR([SDL version >= 1.2.0 not found.])])
+ac_lib_sdl=
+PKG_CHECK_MODULES([SDL], [sdl >= 1.2], [ac_lib_sdl=yes], [AC_MSG_ERROR([SDL version >= 1.2.0 not found.])])
+
+SDL_CFLAGS=`pkg-config  sdl --cflags`
+SDL_LIBS=`pkg-config  sdl --libs`
+
 CFLAGS="$CFLAGS $SDL_CFLAGS"
 CPPFLAGS="$CPPFLAGS $SDL_CFLAGS"
 LIBS="$LIBS $SDL_LIBS"
